[libc][math][c23] Add acospif16() function

amemov opened the pull request:
Addresses #132211  #132754
Part of #95250

amemov committed:
Declared and defined acospif16
- declared acospif16 implementation and header file
- declared test files
- declared in CMakeLists that the op needs to be checked
- defined the op in math.yaml
- declared in the index.rst that the op needs to be checked for float16
- declared an entrypoint for linux x86_64

amemov committed:
Implemented the function and tests
- Implemented acospif16.cpp
- Implemented acospif16_test.cpp for both unit tests and exhaustive tests

TODO:
- Before editing, ninja check-libc was failing the tests. Need to fix

amemov committed:
Update entrypoints.txt

amemov committed:
Upd to resolve merge conflict

amemov committed:
Sync file with upstream main to resolve merge conflicts

amemov committed:
Use upstream version of file to resolve merge conflict

amemov committed:
Added entrypoint of acospif16 to file after syncing with upstream main

amemov committed:
fix

amemov commented:
@Intue - I ended up doing "Synch" of my fork's branch and this change ended up closing my last PR w/o giving me the option to reopen it. 

At least now it should be good in that regard.

lntue commented:
> @lntue - I ended up doing "Synch" of my fork's branch and this change ended up closing my last PR w/o giving me the option to reopen it.
> 
> At least now it should be good in that regard.

A safe-ish way is to fetch llvm repo, rebase llvm/main on your local machine, then force push to your remote branch.

amemov committed:
Addressed the feedback:
- Moved comments about polynomial in line 55 to remove redundancy
- Renamed constexpr array with capital letters
- Added file that was missed in initial commit
- Ran clang-format on files

amemov committed:
Added missing Operation for MPFRUtils.h

lntue commented:
So this is the errors from the buildkite bot:
```
FAILED: projects/libc/src/math/generic/CMakeFiles/libc.src.math.generic.acospif16.__NO_FMA_OPT.dir/acospif16.cpp.o 
...
/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-bt7k7-1/llvm-project/github-pull-requests/libc/src/math/generic/acospif16.cpp:70:17: error: no member named 'polyeval' in namespace '__llvm_libc_21_0_0_git::fputil'
       fputil::polyeval(xsq, POLY_COEFFS[0], POLY_COEFFS[1], POLY_COEFFS[2],
```

amemov committed:
acospif16.h - removed acosf16.h
acospif16.cpp - added missing header
CMakeLists - changed dependencies for acospif16, since it doesn't use ExceptValues like acosf16

amemov commented:
@Intue - I think this time I did not miss anything - triple checked the files I committed and everything seems to be in order. It says for some reason that acospi is undeclared in MPFR - I guess it is what @overmighty in the previous PR I created?
```
FAILED: projects/libc/utils/MPFRWrapper/CMakeFiles/libcMPCommon.dir/MPCommon.cpp.o 

/var/lib/buildkite-agent/builds/linux-56-59b8f5d88-clshd-1/llvm-project/github-pull-requests/libc/utils/MPFRWrapper/MPCommon.cpp:69:3:  error:  use of undeclared identifier 'mpfr_acospi' 
```

lntue commented:
> @lntue - I think this time I did not miss anything - triple checked the files I committed and everything seems to be in order. It says for some reason that acospi is undeclared in MPFR - I guess it is what @overmighty in the previous PR I created?
> 
> ```
> FAILED: projects/libc/utils/MPFRWrapper/CMakeFiles/libcMPCommon.dir/MPCommon.cpp.o 
> 
> /var/lib/buildkite-agent/builds/linux-56-59b8f5d88-clshd-1/llvm-project/github-pull-requests/libc/utils/MPFRWrapper/MPCommon.cpp:69:3:  error:  use of undeclared identifier 'mpfr_acospi' 
> ```

It is because the MPFR version in the build bot machine is older than the latest MPFR, so `mpfr_acospi` is not available.  You will need to add some work around similar to https://github.com/llvm/llvm-project/blob/main/libc/utils/MPFRWrapper/MPCommon.cpp#L393

overmighty commented on libc/include/math.yaml:
```diff
@@ -59,13 +59,6 @@ functions:
     return_type: float
     arguments:
       - type: float
-  - name: asinhf16
```
Why are you removing these 4 functions?

overmighty commented on libc/include/math.yaml:
```diff
@@ -21,19 +21,19 @@ functions:
-    arguments:
-      - type: float
-    name: acoshf16
+  - name: acospif16
```
Nit: sorting: `acoshf16` should be put before `acospif16`.

overmighty commented on libc/src/math/generic/acospif16.cpp:
```diff
@@ -0,0 +1,135 @@
+//===-- Half-precision acospif16(x) function ------------------------------===//
```
Nit:

```suggestion
//===-- Half-precision acospi function ------------------------------------===//
```

overmighty commented on libc/src/math/generic/acospif16.cpp:
```diff
@@ -0,0 +1,135 @@
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception.
+//
+//
```
Nit:

```suggestion
//
```

overmighty commented on libc/test/src/math/smoke/acospif16_test.cpp:
```diff
@@ -0,0 +1,39 @@
+//===-- Unittests for acospif16 -------------------------------------------===//
+//
+//
```
Nit:

```suggestion
//
```

overmighty commented on libc/src/math/generic/acospif16.cpp:
```diff
@@ -0,0 +1,135 @@
+      return x;
+    }
+
+    // 1 < |x| <= +/-inf
```
```suggestion
    // 1 < |x| <= +inf
```

amemov committed:
Addressed style comments:
- Everything pointed out by @overmighty
- Fixed math.yaml to make sure it has all functions from upstream

TODO:
- Modify MPFRWrapper to work with acospi using older version of MPFR

amemov committed:
Addressed some potential style comments:
- Changed order in MPCommon.h and .cpp

Added workaround for older MPFR version
- Support for acospi

Fixed some nit in acospif16.cpp for the comments to look nice

amemov commented:
I implemented a workaround for older versions of MPFR. Right now the acospif16 fails with some values that approach `-1`

Not exactly sure where I messed up - with polynomials in `acospif16.cpp` or with rounding in `MPCommon.cpp` workaround that I wrote

amemov committed:
Changed rounding for mpfr_const_pi in MPCommon.cpp for acospi

amemov committed:
some tweaking in MPCommon.cpp

amemov committed:
Fix for NaN values

amemov committed:
Addressed the feedback for MPCommon.cpp

amemov commented on libc/utils/MPFRWrapper/MPCommon.cpp:
```diff
@@ -70,6 +70,25 @@ MPFRNumber MPFRNumber::acosh() const {
+  MPFRNumber value_pi(0.0, 1280);
+  mpfr_const_pi(value_pi.value, MPFR_RNDN);
+  mpfr_div(result.value, value, value_pi.value, mpfr_rounding);
+  return result;
```
Thanks for the feedback. I was originally going to follow [MPFR's implementation](https://gitlab.inria.fr/mpfr/mpfr/-/blob/master/src/acosu.c) for acospi, but when I checked the examples of workarounds I thought it wouldn't be necessary - so I decided to at least start with something simple to verify that acospif16.cpp works fine for all ranges of values

amemov commented:
@lntue - everything works right now and passes all the checks. 

I do have a question - would the current workaround for mpfr_acospi be acceptable, or should I optimize it further following the [internal implementation of the mpfr_acospi ](https://gitlab.inria.fr/mpfr/mpfr/-/blob/master/src/acosu.c)function by MPFR?

overmighty commented:
Tue is out of office until next week. I tested on an i7-10710U, and I think the test took 2 * 1700 ms with the workaround I suggested. I believe the Buildkite CI runs on hardware that is significantly faster, so it should be fine. I'll try to look into it later today.

overmighty commented:
It turns out that the Buildkite CI's hardware is not as fast as I thought, we can see in the latest run's log that it took 2 * 3 seconds to run `LlvmLibcAcospif16Test`. But seemingly noone sees it as an actual problem.

overmighty commented on libc/src/math/CMakeLists.txt:
```diff
@@ -43,6 +43,7 @@ endfunction()
 add_math_entrypoint_object(acos)
 add_math_entrypoint_object(acosf)
 add_math_entrypoint_object(acosf16)
+add_math_entrypoint_object(acospif16)
 
 add_math_entrypoint_object(acosh)
 add_math_entrypoint_object(acoshf)
```
Nit: should be grouped/sorted like this:

```cmake
add_math_entrypoint_object(acosh)
add_math_entrypoint_object(acoshf)
add_math_entrypoint_object(acoshf16)

add_math_entrypoint_object(acospif16)

add_math_entrypoint_object(asin)
add_math_entrypoint_object(asinf)
add_math_entrypoint_object(asinf16)
```

overmighty reviewed (approved):
LGTM besides the last nitpick I posted above. (I accidentally clicked "Add single comment" instead of "Start a review.")

amemov committed:
Reordered 'acos' entrypoint in alphabetical order

overmighty commented on libc/src/math/CMakeLists.txt:
```diff
@@ -41,13 +41,14 @@ function(add_math_entrypoint_object name)
 add_math_entrypoint_object(acos)
-add_math_entrypoint_object(acosf)
-add_math_entrypoint_object(acosf16)
-
 add_math_entrypoint_object(acosh)
+add_math_entrypoint_object(acosf)
 add_math_entrypoint_object(acoshf)
+add_math_entrypoint_object(acosf16)
 add_math_entrypoint_object(acoshf16)
 
+add_math_entrypoint_object(acospif16)
+ 
 add_math_entrypoint_object(asin)
 add_math_entrypoint_object(asinf)
 add_math_entrypoint_object(asinf16)
```
I meant putting `acospif16` between the `acosh*` and `asin*` functions, like this:

```suggestion
add_math_entrypoint_object(acos)
add_math_entrypoint_object(acosf)
add_math_entrypoint_object(acosf16)

add_math_entrypoint_object(acosh)
add_math_entrypoint_object(acoshf)
add_math_entrypoint_object(acoshf16)

add_math_entrypoint_object(acospif16)

add_math_entrypoint_object(asin)
add_math_entrypoint_object(asinf)
add_math_entrypoint_object(asinf16)
```

amemov replied:
I think it should be good now

amemov committed:
Fixed the ordering
